packages/lib/services/survey.ts

```diff
@@ -78,183 +82,245 @@ export const preloadSurveyWithAnalytics = (surveyId: string) => {
+        }
+
+        throw error;
       }
 
-      throw error;
-    }
+      if (!surveyPrisma) {
+        throw new ResourceNotFoundError("Survey", surveyId);
+      }
 
-    if (!surveyPrisma) {
-      throw new ResourceNotFoundError("Survey", surveyId);
-    }
+      let { _count, displays, ...surveyPrismaFields } = surveyPrisma;
 
-    let { _count, displays, ...surveyPrismaFields } = surveyPrisma;
-
-    const numDisplays = displays.length;
-    const numDisplaysResponded = displays.filter((item) => item.status === "responded").length;
-    const numResponses = _count.responses;
-    // responseRate, rounded to 2 decimal places
-    const responseRate = numDisplays ? Math.round((numDisplaysResponded / numDisplays) * 100) / 100 : 0;
-
-    const transformedSurvey = {
-      ...surveyPrismaFields,
-      triggers: surveyPrismaFields.triggers.map((trigger) => trigger.eventClass),
-      analytics: {
-        numDisplays,
-        responseRate,
-        numResponses,
-      },
-    };
+      const numDisplays = displays.length;
+      const numDisplaysResponded = displays.filter((item) => item.status === "responded").length;
+      const numResponses = _count.responses;
+      // responseRate, rounded to 2 decimal places
+      const responseRate = numDisplays ? Math.round((numDisplaysResponded / numDisplays) * 100) / 100 : 0;
 
-    try {
-      const survey = ZSurveyWithAnalytics.parse(transformedSurvey);
-      return survey;
-    } catch (error) {
-      console.log(error);
-      if (error instanceof z.ZodError) {
-        console.error(JSON.stringify(error.errors, null, 2)); // log the detailed error information
-      }
-      throw new ValidationError("Data validation of survey failed");
-    }
-  }
-);
+      const transformedSurvey = {
+        ...surveyPrismaFields,
+        triggers: surveyPrismaFields.triggers.map((trigger) => trigger.eventClass),
+        analytics: {
+          numDisplays,
+          responseRate,
+          numResponses,
+        },
+      };
 
-export const getSurvey = cache(async (surveyId: string): Promise<TSurvey | null> => {
-  validateInputs([surveyId, ZId]);
-  let surveyPrisma;
-  try {
-    surveyPrisma = await prisma.survey.findUnique({
-      where: {
-        id: surveyId,
-      },
-      select: selectSurvey,
-    });
-  } catch (error) {
-    if (error instanceof Prisma.PrismaClientKnownRequestError) {
-      throw new DatabaseError("Database operation failed");
+      try {
+        const survey = ZSurveyWithAnalytics.parse(transformedSurvey);
+        return survey;
+      } catch (error) {
+        console.log(error);
+        if (error instanceof z.ZodError) {
+          console.error(JSON.stringify(error.errors, null, 2)); // log the detailed error information
+        }
+        throw new ValidationError("Data validation of survey failed");
+      }
+    },
+    [getSurveyWithAnalyticsCacheTag(surveyId)],
+    {
+      tags: [getSurveyWithAnalyticsCacheTag(surveyId)],
```

Comment:
SurveysWithAnalytics also depend on the displays and the responses and thus the cache should also gets invalidated when a display or a response changes (gets added, gets deleted, gets changed).
I think in this cache it doesn't make sense to have only one cache tag, but have standardized cache tags for e.g. `environments-${envId}-people` (exported from the people service) or `environments-${envId}-displays` exported from the displays service. What do you think?

Reply:
In this case we also can't have the cache key the same as the tag. Also I think pretty soon it will happen that we have multiple cached functions calling the same ressource (e.g. `getProduct` & `getProductByEnvironmentId`).
Should / Can these functions have the same caching key?
(If they can, I think we should name them by the ressource (e.g.`products-${productId}`))
(In this case this would also be `SurveysWithAnalytics-${surveyId}`? Always not sure if we name them from our types or the database models. I think for keys I would go by types as they are more unique, but for tags just say: they are surveys and should change when surveys change.)

Reply:
I hope this can stay consistent :-D